introspection-engine/connectors/sql-introspection-connector/src/calculate_datamodel.rs

```diff
@@ -107,16 +160,100 @@ pub fn calculate_datamodel(
         previous_schema: ctx.previous_schema(),
         warnings: &mut warnings,
         existing_enums: &existing_enums,
+        existing_models: &existing_models,
+        existing_scalar_fields: &existing_scalar_fields,
+        existing_relations: &existing_relations,
+        existing_m2m_relations: &existing_m2m_relations,
     };
 
     let (version, data_model, is_empty) = introspect(&mut context)?;
 
-    debug!("Done calculating datamodel.");
-
     Ok(IntrospectionResult {
         data_model,
         is_empty,
         warnings,
         version,
     })
 }
+
+fn match_existing_models(schema: &sql::SqlSchema, ctx: &IntrospectionContext) -> HashMap<sql::TableId, ast::ModelId> {
+    ctx.previous_schema()
+        .db
+        .walk_models()
+        .filter_map(|model| {
+            schema
+                .find_table(model.database_name())
+                .map(|sql_id| (sql_id, model.id))
+        })
+        .collect()
+}
+
+fn match_existing_scalar_fields(
+    existing_models: &HashMap<sql::TableId, ast::ModelId>,
+    schema: &sql::SqlSchema,
+    ctx: &IntrospectionContext,
+) -> HashMap<sql::ColumnId, (ast::ModelId, ast::FieldId)> {
+    schema
+        .walk_columns()
+        .filter_map(|col| {
+            let model_id = existing_models.get(&col.table().id)?;
+            let field_id = ctx
+                .previous_schema()
+                .db
+                .walk(*model_id)
+                .scalar_fields()
+                .find(|field| field.database_name() == col.name())
+                .map(|field| field.field_id())?;
+            Some((col.id, (*model_id, field_id)))
+        })
+        .collect()
+}
+
+fn match_existing_inline_relations(
+    existing_models: &HashMap<sql::TableId, ast::ModelId>,
+    schema: &sql::SqlSchema,
+    ctx: &IntrospectionContext,
+) -> HashMap<sql::ForeignKeyId, RelationId> {
+    schema
+        .walk_foreign_keys()
+        .filter_map(|fk| {
+            let referencing_model = *existing_models.get(&fk.table().id)?;
+            ctx.previous_schema()
+                .db
+                .walk(referencing_model)
+                .relations_from()
+                .filter_map(|rel| rel.refine().as_inline())
+                .find(|relation| {
+                    let referencing_fields = if let Some(fields) = relation.referencing_fields() {
+                        fields
+                    } else {
+                        return false;
+                    };
+                    let referencing_columns = fk.constrained_columns();
+                    referencing_fields.len() == referencing_columns.len()
+                        && referencing_fields
+                            .zip(referencing_columns)
+                            .all(|(field, col)| field.database_name() == col.name())
+                })
+                .map(|relation| (fk.id, relation.relation_id()))
+        })
```

Comment:
This one I'm a bit ambivalent about: if we `filter_map()` we don't have the relation anymore, but we need it on line 238. So it would have to be a `.filter_map(|relation| relation.fields().map(|fields| (relation, fields))`. I'll see if that fits nicely on one line.